src/v/storage/kvstore.cc

```diff
@@ -460,6 +460,7 @@ ss::future<> kvstore::load_snapshot_from_reader(snapshot_reader& reader) {
           batch.header().header_crc));
     }
 
+    _db.reserve(batch.header().last_offset() - batch.header().base_offset);
```

Comment:
if this was found to help in practice, then it might be an omen that _db is growing to contain enough entries that it's underlying contiguous vector allocation is oom sensitive, in which case switching to absl btree_map may be warranted. looks like each entry is roughly 80 bytes or so

Reply:
Yeah, we are in a borderline case here, it can get into the 100s of kB.

Unfortunately record batch also contains a contiguous vector of records, so really getting rid of the O(n_entries) contiguous allocations requires either making record batch a fragmented vector, or reworking kvstore snapshots to use multiple batches.

Reply:
hmm, not entirely sure we are talking about the same thing.

there are two contiguous allocations to consider. the vector backing _db where the only consideration is something like the static `sizeof(bytes) + sizeof(iobuf)` * dynamic `num_entries`. btree_map is like the fragmented_vector analog for associative containers.

> Unfortunately record batch also contains a contiguous vector of records

the second is the value: an iobuf. so it's already heap allocated for bulk data and fragmented.

Reply:
sgtm